realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -492,20 +494,31 @@ synchronized void invokeWithLock(Callback0 callback) {
      * @param configuration configuration object for Realm instance.
      * @throws RealmFileException if copying the file fails.
      */
-    private static void copyAssetFileIfNeeded(RealmConfiguration configuration) {
-        if (configuration.hasAssetFile()) {
-            File realmFile = new File(configuration.getRealmDirectory(), configuration.getRealmFileName());
-
-            copyFileIfNeeded(configuration.getAssetFilePath(), realmFile);
-        }
+    private static void copyAssetFileIfNeeded(final RealmConfiguration configuration) {
+        final File realmFileFromAsset = configuration.hasAssetFile() ?
+                new File(configuration.getRealmDirectory(), configuration.getRealmFileName())
+                : null;
+        final String syncServerCertificateAssetName = ObjectServerFacade.getFacade(
+                configuration.isSyncConfiguration()).getSyncServerCertificateAssetName(configuration);
+
+        if (realmFileFromAsset!= null || !Util.isEmptyString(syncServerCertificateAssetName)) {
+            OsObjectStore.callWithLock(configuration, new Runnable() {
```

Comment:
why copying the asset using the lock? this is part of the initialization ... I rather have the entire initialization done within a lock (rather than parts of it) ...

Reply:
the problem is, there could be another process/sync client opened the Realm at the same time. so the asset file could be copied already.
When core is holding the lock file, the realm file is not necessarily to be there.

it is like below:

core holding the lock file ->
check if file existence ->
copy the file ->
core release the lock